src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1350,9 +1350,148 @@ text \<open>
   commutativity rules.
 *)
 
+lemma repeated_receive_evolution:
+  assumes "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>\<alpha>\<rparr> Q"
+  obtains n and X
+  where "\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X"
+    and "Q = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
+  using assms
+  by
+    (subst (asm) (2) repeated_receive_proper_def)
+    (rotate_tac, induction \<alpha> "A \<triangleright> x. (\<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" Q rule: synchronous_transition.induct,
+     fastforce intro: repeated_receive_proper_def)
+
```

Comment:
This proof already features key ingredients like the use of `subst` and the use of a properly instantiated elimination rule (through the use of `induction`), but it can be further simplified.

The first thing to observe is that we actually don’t need induction, but only case distinction (going down into cases one time instead of arbitrarily many times). This suggests to replace the proof method expression `induction α "A ▹ x. (𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)" Q rule: synchronous_transition.induct` by `cases α "A ▹ x. (𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)" Q rule: synchronous_transition.cases`.

A side effect of using `cases` instead of `induction` is that the `rotate_tac` isn’t needed anymore. The `induction` method expects the application of the inductive predicate as the first premise or chained fact, but `cases` apparently just creates one subgoal for each premise, with the subgoal created for the application of the predicate being so trivial that it can be solved by the `assumption` method.

Interestingly, also the other subgoal created by `cases` is trivial. As a result, something like the `fastforce` invocation above isn’t needed anymore, but a simple `assumption` also suffices here.

Since Isar applies proof by assumption automatically, the proof method expressions after `by` can be reduced to just the `subst` and the `cases` invocation, leading to the following proof:
```isabelle
using assms
by
  (subst (asm) (2) repeated_receive_proper_def)
  (cases α "A ▹ x. (𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)" Q rule: synchronous_transition.cases)
```

However, when using `by`, the first proof method expression should typically reduce the goal to some subgoals referring to different cases, which the second proof method expression should solve. Therefore, an application of `cases` like the one above should really go into the first proof method expression. Therefore, I suggest to bundle both proof method expressions into a single one.

However, doing only that doesn’t work, probably because it results in `assms` being passed as a chained fact to the invocation of `cases` as well. A fix is to pass it only to the `subst` invocation using the `use` method. This leads to the following proof, which I suggest to adopt:
```isabelle
by
  (
    use assms in ‹subst (asm) (2) repeated_receive_proper_def›,
    cases α "A ▹ x. (𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)" Q rule: synchronous_transition.cases
  )
```

Reply:
Thanks a lot for your detailed explanation, it was very useful. The enhancement is included in 5958b601995aaf54e72588a5a6959f66e665bd7f, please verify.

Reply:
Looks good to me. :slightly_smiling_face: